src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs

```diff
@@ -86,39 +92,84 @@ public static int GetMaxEscapedLength(int textLength, int firstIndexToEscape)
             return firstIndexToEscape + JsonConstants.MaxExpansionFactorWhileEscaping * (textLength - firstIndexToEscape);
         }
 
-        public static void EscapeString(ReadOnlySpan<byte> value, Span<byte> destination, int indexOfFirstByteToEscape, out int written)
+        public static void EscapeString(ReadOnlySpan<byte> value, Span<byte> destination, int indexOfFirstByteToEscape, JavaScriptEncoder encoder, out int written)
         {
             Debug.Assert(indexOfFirstByteToEscape >= 0 && indexOfFirstByteToEscape < value.Length);
 
             value.Slice(0, indexOfFirstByteToEscape).CopyTo(destination);
             written = indexOfFirstByteToEscape;
             int consumed = indexOfFirstByteToEscape;
 
-            while (consumed < value.Length)
+            if (encoder != null)
             {
-                byte val = value[consumed];
-                if (NeedsEscaping(val))
+                OperationStatus result = encoder.EncodeUtf8(
+                    value.Slice(consumed), destination.Slice(written), out int encoderBytesConsumed, out int encoderBytesWritten);
+
+                Debug.Assert(result != OperationStatus.DestinationTooSmall);
+                Debug.Assert(result != OperationStatus.NeedMoreData);
+                Debug.Assert(encoderBytesConsumed == value.Length - consumed);
+
+                if (result != OperationStatus.Done)
                 {
-                    consumed += EscapeNextBytes(value.Slice(consumed), destination, ref written);
+                    ThrowHelper.ThrowArgumentException_InvalidUTF8(value.Slice(encoderBytesWritten));
                 }
-                else
+
+                written += encoderBytesWritten;
+            }
+            else
+            {
+                while (consumed < value.Length)
```

Comment:
This is strictly for performance. I'll add a comment. The logic is duplicated, but is ~40% faster than calling into the default encoder.

An alternative is to have our own "default" encoder (deriving from JavaScriptEncoder) which then could have this code in it (but still duplicated logic). This will not be faster, and may be slightly slower (due to virtual call and slicing), but would keep the logic together at least.